src/Autofac/Util/Cache/TypeAssemblyReferenceProvider.cs

```diff
@@ -68,6 +68,11 @@ private static void PopulateAllReferencedAssemblies(Type inputType, HashSet<Asse
 
         foreach (var genericArgumentType in genericArguments)
         {
+            if (holdingSet.Any(a => a.DefinedTypes.Contains(genericArgumentType)))
```

Comment:
For performance reasons, if this only overflows in a tight loop where we're self-referencing, can we please pass into PopulateAllReferencedAssemblies a "HashSet<Type> activeWorkingSet" and insert there each time we add a type, then check that active working set instead? This should result in a simple O(1) check for the type having already been seen in this tree.

Reply:
Sure thing.  Changes have been committed.